lib/listen/record.rb

```diff
@@ -75,31 +63,36 @@ def build
 
     private
 
+    # TODO: refactor/refactor out
+    def add_dir(dir, rel_path, children)
+      rel_path = '.' if [nil, '', '.'].include? rel_path
+      dirname, basename = Pathname(rel_path).split.map(&:to_s)
+      basename = '.' if [nil, '', '.'].include? basename
+      if [nil, '', '.'].include?(dirname) && basename == '.'
+        @paths[dir.to_s]['.'] = Hash[children.map { |child| [child, {} ] } ]
+      else
+        @paths[dir.to_s][rel_path] ||= {}
```

Comment:
i don't think the children here is needed because they'll be iterated during `_fast_build` anyway.
Also `Directory::scan` would touch each entry populating new directories and their contents.

Also might be a mistake here either way as it only does the children for the root folder?

Lastly the bug with the moving of a folder not showing the old folder in the removed list is still present as the folder itself is not added to its parent directory's hash. This was missing from my patch too. I did a test where in `add_dir` it adds the directory hash as well as adding itself to the hash of the parent directory and that seemed to work.

So in essencing adding directory takes two actions:
1. Set up hash for contents
2. Add it to its parent directory's contents hash